src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -1000,603 +946,315 @@ private ImmutableArray<Symbol> CreateMembers()
             {
-                var member = members[i];
-                switch (member.Kind)
+                var fields = ArrayBuilder<Symbol>.GetInstance();
+                foreach (var member in members)
                 {
-                    case SymbolKind.Method:
-                        map.Add(((MethodSymbol)member).TupleUnderlyingMethod.OriginalDefinition, member);
-                        break;
-
-                    case SymbolKind.Field:
-                        var tupleUnderlyingField = ((FieldSymbol)member).TupleUnderlyingField;
-                        if ((object)tupleUnderlyingField != null)
-                        {
-                            map[tupleUnderlyingField.OriginalDefinition] = member;
-                        }
-                        break;
-
-                    case SymbolKind.Property:
-                        map.Add(((PropertySymbol)member).TupleUnderlyingProperty.OriginalDefinition, member);
-                        break;
-
-                    case SymbolKind.Event:
-                        var underlyingEvent = ((EventSymbol)member).TupleUnderlyingEvent;
-                        var underlyingAssociatedField = underlyingEvent.AssociatedField;
-                        // The field is not part of the members list
-                        if ((object)underlyingAssociatedField != null)
-                        {
-                            Debug.Assert((object)underlyingAssociatedField.ContainingSymbol == _underlyingType);
-                            Debug.Assert(_underlyingType.GetMembers(underlyingAssociatedField.Name).IndexOf(underlyingAssociatedField) < 0);
-                            map.Add(underlyingAssociatedField.OriginalDefinition, new TupleFieldSymbol(this, underlyingAssociatedField, -i - 1));
-                        }
-
-                        map.Add(underlyingEvent.OriginalDefinition, member);
-                        break;
-
-                    default:
-                        throw ExceptionUtilities.UnexpectedValue(member.Kind);
+                    if (member is TupleVirtualElementFieldSymbol)
+                    {
+                        continue;
+                    }
+                    else if (member is TupleElementFieldSymbol tupleField)
+                    {
+                        fields.Add(tupleField.UnderlyingField.OriginalDefinition);
+                    }
+                    else if (member is FieldSymbol field)
+                    {
+                        fields.Add(field.OriginalDefinition);
+                    }
                 }
-            }
 
-            return map;
+                Debug.Assert(fields.All(f => f is object));
+                return fields.ToImmutableAndFree();
+            }
         }
 
-        public TMember GetTupleMemberSymbolForUnderlyingMember<TMember>(TMember underlyingMemberOpt) where TMember : Symbol
+        private TypeSymbol MergeTupleNames(TypeSymbol other, NamedTypeSymbol mergedType)
```

Comment:
>TypeSymbol MergeTupleNames(TypeSymbol other [](start = 16, length = 43)

Consider changing signature to `NamedTypeSymbol MergeTupleNames(NamedTypeSymbol ...`.